tools/validators/instance_validator/validate/telemetry_validator.py

```diff
@@ -0,0 +1,128 @@
+specfied timeout duration is reached."""
+
+import threading
+
+from validate import telemetry
+from validate import telemetry_error
+
+DEVICE_ID = "deviceId"
+TRANSLATION = "translation"
+STATES = "states"
+UNITS = "unit_values"
+
+class TelemetryValidator(object):
+  """Validates telemetry messages against a building config file.
+
+  Args:
+    entities: parsed entities from building config
+    timeout: validation timeout duration in seconds
+    callback: callback function to be called when the validation finishes
+  """
+
+  def __init__(self, entities, timeout, callback):
+    super().__init__()
+    self.entities = entities
+    self.timeout = timeout
+    self.callback = callback
+    self.validated_entities = {}
+    self.validation_errors = []
+
+  def StartTimer(self):
+    """Starts the validation timeout timer."""
+    threading.Timer(self.timeout, lambda: self.callback(self)).start()
+
+  def AllEntitiesValidated(self):
+    """Returns true if a message was received for every entity."""
+    return len(self.entities) == len(self.validated_entities)
+
+  def CallbackIfCompleted(self):
+    """Checks if all entities have been validated, and calls the callback."""
+    if self.AllEntitiesValidated():
+      self.callback(self)
+
+  def AddError(self, error):
+    """Adds a validation error."""
+    self.validation_errors.append(error)
+
+  def ValidateMessage(self, message):
+    """Validates a telemetry message."""
+    t = telemetry.Telemetry(message)
+    entity_name = t.attributes[DEVICE_ID]
+
+    if entity_name not in self.entities:
+      self.AddError(
+        telemetry_error.TelemetryError(
+          entity_name, None, "Unknown entity"))
+      message.ack()
+      return
+
+    if entity_name in self.validated_entities:
+      # Already validated telemetry for this entity,
+      # so the message can be skipped.
+      return
+    self.validated_entities[entity_name] = True
+
+    entity = self.entities[entity_name]
+    for point_name, point_config in entity[TRANSLATION].items():
+      if point_name not in t.points.keys():
+        self.AddError(
+          telemetry_error.TelemetryError(
+            entity_name, point_name, "Missing point"))
+        continue
+
+      point = t.points[point_name]
+      pv = point.present_value
+      if pv is None:
+        self.AddError(
+          telemetry_error.TelemetryError(
+            entity_name, point_name, "Missing present value"))
+        continue
+
+      has_states = STATES in point_config
+      has_units = UNITS in point_config
+
+      if has_states:
+        states = point_config[STATES]
+        if pv not in states.values():
+          self.AddError(
+            telemetry_error.TelemetryError(
+              entity_name, point_name, "Invalid state: {}".format(pv)))
+          continue
+
+      if has_units and not self.ValueIsNumeric(pv):
+        handle_bool = lambda x: str(x).lower() if isinstance(x, bool) else x
```

Comment:
why lower for bool but verbatim for others? (just curious)

Reply:
The intent was to deal with a) boolean values being parseable as numbers in Python, and then b) the discrepancy between the lowercase 'false' in yaml and the capitalized 'False' in Python. I thought about it again and decided that the case discrepancy in error reporting isn't a big deal, and it's not guaranteed that the MSI would lowercase 'false', e.g. they might write 'FALSE' (which I'm pretty sure still parses to Python 'False'), in which case the error message wouldn't have the correct case anyways.

Anyway, I've removed this special case.